File: src/recall/fleetwatch.py
"""Report recall's health to fleetwatch — the fleet's monitoring platform.

fleetwatch is push-based: a producer POSTs verdict-shaped reports and it keeps their
history (see its README, "The report contract"). Two properties make it the right place
for this, and they are why nothing here needs to send mail or run a daemon that watches
a daemon:

* **A dead producer is a failure, not a silence.** A report declares its own cadence
  (`interval_s`); fleetwatch renders a producer that has stopped reporting as `Silent` —
  red. So this need not detect the case where the *Mac itself* dies: not reporting IS
  the report. A monitor that only speaks when it is well is no monitor at all.
* The verdicts already show up where the rest of the fleet's health does, which is where
  a fault will actually be seen.

The ingest token is read from `~/.config/fleetwatch/token` and never leaves this machine
— it is not in the repo, not in the report, and not logged.
"""

from __future__ import annotations

import json
import os
import secrets
import urllib.error
import urllib.request
from collections.abc import Sequence
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path

from recall.health import Check

DEFAULT_URL = "https://fleetwatch.xinutec.org/api/reports"
TOKEN_FILE = Path.home() / ".config" / "fleetwatch" / "token"
# One producer for the whole of recall: is it recording, are its agents up, is the
# archive mirrored. Kept as one collector so a single fleetwatch tile answers "is
# recall alright?" — the question actually being asked.
COLLECTOR = "recall"
# Declared cadence. fleetwatch turns this into staleness: report less often than
# this and the tile goes amber, stop entirely and it goes red. It must match the
# launchd agent's StartInterval, or a healthy producer is reported as late.
INTERVAL_S = 300
_SCHEMA = 1
_CROCKFORD = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"  # no I, L, O, U — a ULID alphabet


def mint_ulid(now: datetime, randomness: bytes) -> str:
    """A ULID: 48 bits of millisecond timestamp, then 80 bits of randomness, in
    Crockford base32. fleetwatch uses it as the idempotency key and rejects anything
    that is not one (422), so it is minted here rather than depending on a library for
    twelve lines. Pure, so it is tested against a known vector.
    """
    if len(randomness) != 10:  # noqa: PLR2004 - the ULID spec's 80 bits
        raise ValueError("a ULID needs exactly 80 bits of randomness")
    value = (int(now.timestamp() * 1000) << 80) | int.from_bytes(randomness, "big")
    return "".join(_CROCKFORD[(value >> shift) & 0x1F] for shift in range(125, -1, -5))


@dataclass(frozen=True)
class Report:
    """One fleetwatch report. `source` is deliberately absent: fleetwatch stamps it from
    the ingest token, so a producer can only ever write as itself."""

    id: str
    collected_at: datetime
    checks: tuple[Check, ...]
    duration_ms: int | None = None

    def payload(self) -> dict[str, object]:
        return {
            "schema": _SCHEMA,
            "id": self.id,
            "collector": COLLECTOR,
            "collected_at": self.collected_at.isoformat(),
            "duration_ms": self.duration_ms,
            "interval_s": INTERVAL_S,
            "checks": [
                {
                    "section": check.section,
                    "label": check.label,
                    "verdict": check.verdict,
                    "observed": check.observed,
                    "expected": check.expected,
                    "value": check.value,
                    "unit": check.unit,
                }
                for check in self.checks
            ],
        }


def build_report(
    checks: Sequence[Check],
    *,
    now: datetime,
    randomness: bytes | None = None,
    duration_ms: int | None = None,
) -> Report:
    return Report(
        id=mint_ulid(
            now, randomness if randomness is not None else secrets.token_bytes(10)
        ),
        collected_at=now,
        checks=tuple(checks),
        duration_ms=duration_ms,
    )


def read_token(path: Path = TOKEN_FILE) -> str | None:
    """The ingest token, from the environment or the file the fleet's secret.sh writes.

    None if there is none — this machine is simply not a producer yet, which is a thing
    to say plainly, not to crash on.
    """
    from_env = os.environ.get("RECALL_FLEETWATCH_TOKEN")
    if from_env:
        return from_env.strip()
    try:
        token = path.read_text().strip()
    except OSError:
        return None
    return token or None


def post_report(report: Report, *, token: str, url: str = DEFAULT_URL) -> int:
    """POST one report. Returns the HTTP status (201 stored, 200 duplicate).

    Raises on a transport failure — the caller decides whether a fleetwatch it cannot
    reach is worth failing over. It is not: the recording is what matters, and an
    unreachable monitor already shows itself as stale at the other end.
    """
    request = urllib.request.Request(
        url,
        data=json.dumps(report.payload()).encode(),
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {token}",
        },
        method="POST",
    )
    with urllib.request.urlopen(request, timeout=15) as response:
        return int(response.status)

File: doctor/src/fleetwatch.rs
//! Report recall's health to fleetwatch, the fleet's monitoring platform.
//!
//! fleetwatch is push-based and keeps the history (its README, "The report
//! contract"). A report declares its cadence (`interval_s`) and a producer that
//! stops reporting renders red, so a dead Mac needs no detector of its own.
//!
//! The ingest token comes from `RECALL_FLEETWATCH_TOKEN` or
//! `~/.config/fleetwatch/token`, and is never put in the report or logged.

use crate::check::Check;
use chrono::{DateTime, Utc};
use std::path::Path;
use std::time::Duration;

pub const DEFAULT_URL: &str = "https://fleetwatch.xinutec.org/api/reports";
/// One collector for the whole of recall, so a single fleetwatch tile answers
/// "is recall alright?".
const COLLECTOR: &str = "recall";
/// Declared cadence. fleetwatch turns this into staleness: report less often
/// than this and the tile goes amber, stop entirely and it goes red. It must
/// match the launchd agent's `StartInterval`, or a healthy producer is reported
/// as late.
const INTERVAL_S: u32 = 300;
const SCHEMA: u32 = 1;
/// A ULID alphabet — no I, L, O, U.
const CROCKFORD: &[u8; 32] = b"0123456789ABCDEFGHJKMNPQRSTVWXYZ";

/// A ULID: 48 bits of millisecond timestamp, then 80 bits of randomness, in
/// Crockford base32. fleetwatch uses it as the idempotency key and rejects
/// anything that is not one (422), so it is minted here rather than depending
/// on a crate for a dozen lines. Pure, so it is tested against a known vector.
pub fn mint_ulid(now: DateTime<Utc>, randomness: [u8; 10]) -> String {
    let value = ((now.timestamp_millis() as u128) << 80)
        | u128::from_be_bytes({
            let mut wide = [0u8; 16];
            wide[6..].copy_from_slice(&randomness);
            wide
        });
    (0..26)
        .map(|i| {
            let shift = 125 - i * 5;
            CROCKFORD[((value >> shift) & 0x1F) as usize] as char
        })
        .collect()
}

/// 80 bits from the kernel. `/dev/urandom` rather than a crate: this is the
/// only randomness in the binary and it is an idempotency key, not a secret.
fn randomness() -> [u8; 10] {
    use std::io::Read as _;
    let mut bytes = [0u8; 10];
    if let Ok(mut file) = std::fs::File::open("/dev/urandom") {
        let _ = file.read_exact(&mut bytes);
    }
    bytes
}

/// The report body. `source` is deliberately absent: fleetwatch stamps it from
/// the ingest token, so a producer can only ever write as itself.
pub fn payload(
    checks: &[Check],
    now: DateTime<Utc>,
    duration_ms: Option<u64>,
) -> serde_json::Value {
    serde_json::json!({
        "schema": SCHEMA,
        "id": mint_ulid(now, randomness()),
        "collector": COLLECTOR,
        "collected_at": now.to_rfc3339_opts(chrono::SecondsFormat::Micros, false),
        "duration_ms": duration_ms,
        "interval_s": INTERVAL_S,
        "checks": checks,
    })
}

/// The ingest token, from the environment or the file the fleet's `secret.sh`
/// writes. `None` if there is none — this machine is simply not a producer yet,
/// which is a thing to say plainly, not to crash on.
/// `from_env` is passed in rather than read here: which of the two sources wins
/// is a decision worth testing, and setting an environment variable is `unsafe`
/// in this edition — a test that mutated process-global state under a threaded
/// runner would be the wrong way to reach it.
pub fn read_token(home: &Path, from_env: Option<&str>) -> Option<String> {
    let env = from_env.map(str::trim).filter(|t| !t.is_empty());
    if let Some(token) = env {
        return Some(token.to_owned());
    }
    let path = home.join(".config").join("fleetwatch").join("token");
    let token = std::fs::read_to_string(path).ok()?;
    let token = token.trim();
    (!token.is_empty()).then(|| token.to_owned())
}

/// POST one report. Returns the HTTP status (201 stored, 200 duplicate).
///
/// Fails on a transport error — the caller decides whether a fleetwatch it
/// cannot reach is worth failing over. It is not: the recording is what
/// matters, and an unreachable monitor already shows itself as stale at the
/// other end.
pub fn post(body: &serde_json::Value, token: &str, url: &str) -> Result<u16, Box<ureq::Error>> {
    let agent = ureq::AgentBuilder::new()
        .timeout(Duration::from_secs(15))
        .build();
    let response = agent
        .post(url)
        .set("Content-Type", "application/json")
        .set("Authorization", &format!("Bearer {token}"))
        .send_json(body)
        .map_err(Box::new)?;
    Ok(response.status())
}
